### PR TITLE
Fix a use after free for Managed Data BP

### PR DESCRIPTION
**Pull request by chuckries**
ShimProxyCallback::DataBreakpoint::DataBreakpointEvent was holding onto a bare
BYTE* for the CONTEXT rather than copying the buffer and taking ownership. Due to
lifetime, this resulted in a use after free. Apparently in retail code we got lucky
and this worked enough of the time that we never noticed it.

### CODE REVIEW

**Comment by chuckries** [comment]
@cshung @noahfalk please take a look.

**Comment by chuckries** [comment]
@tommcdon This should really go into preview 1. Is it too late?

File: src/debug/di/shimcallback.cpp [hunk 1]
```diff
@@ -1402,10 +1402,22 @@ HRESULT ShimProxyCallback::DataBreakpoint(ICorDebugProcess* pProcess, ICorDebugT
         {
             this->m_pProcess.Assign(pProcess);
             this->m_pThread.Assign(pThread);
-            this->m_pContext = pContext;
+
+            this->m_pContext = new BYTE[contextSize];
```
Comment (cshung) [inline src/debug/di/shimcallback.cpp]: I am wondering about OOM handling here, we can't return anything in a constructor.
How about using a CONTEXT field instead?

File: src/debug/di/shimcallback.cpp [hunk 2]
```diff
@@ -1402,10 +1402,22 @@ HRESULT ShimProxyCallback::DataBreakpoint(ICorDebugProcess* pProcess, ICorDebugT
...
+            memcpy(this->m_pContext, pContext, contextSize);
+
             this->m_contextSize = contextSize;
         }
 
+        ~DataBreakpointEvent()
```
Comment (cshung) [inline src/debug/di/shimcallback.cpp]: This can all go away if we simply use a CONTEXT field.

File: src/debug/di/shimcallback.cpp (hunk 1 above)
Comment (chuckries) [reply src/debug/di/shimcallback.cpp]: should I assert that contextSize == sizeof(CONTEXT)?

**Pull request merged**